x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -482,6 +482,17 @@ private Map<String, Object> buildUserMetadata(JWTClaimsSet claimsSet) {
         return Map.copyOf(metadata);
     }
 
+    // We construct the token principal as a function of the JWT realm configuration. We also short circuit the extraction of the
+    // token principal while we iterate through the realms. For realms like the file realm this is not an issue since there is only
+    // one file realm. For realms like LDAP this is also not an issue since the token principal is identical across all realms regardless
+    // of how the realm is configured. However, for realms like JWT (and PKI realm) where the token principal is a function of the
+    // realm configuration AND multiple realms of that type can exist this can be an issue. This is an issue because realm1 might
+    // result in the token principal "abc", but realm2 (same JWT) might result in the token principal as "xyz". Since we short circuit the
+    // extraction of the token principal (i.e. use the first one that does not error) then the same JWT token can result in a
+    // token principal of either "abc" or "xyz" depending on which came first. To compound issue the order in which realms are evaluated is
+    // effectively non-deterministic due to smart re-ordering and caching. This means that we can not rely on the value calculated here
```

Comment:
Realm smart ordering does not affect the token extraction process because the smart ordering requires the token principal to be extracted first. So for the same JWT, the token principal should keep being the same until realm YAML configuration changes (and node restart), e.g. re-order the realms, or remove the higher order realm that previously extracted the token principal but was not the one that authenticated it, or adding another higher order realm that happen to extract the token principal differently.

Reply:
I think if the cache expires and subsequent attempts to build the token principal fail in the first realm where it formerly worked then the token principal can change without a restart or explicit reorder. That is probably an extreme edge case which requires a lot of conditions and wasn't the original intent calling this out and it does distract from the larger point... will remove.